rp2040-hal/src/gpio/func.rs

```diff
@@ -0,0 +1,185 @@
+use core::marker::PhantomData;
+
+use paste::paste;
+
+use super::pin::DynBankId;
+
+pub(crate) mod func_sealed {
+    use super::DynFunction;
+
+    pub trait Function {
+        fn from(f: DynFunction) -> Self;
+        fn as_dyn(&self) -> DynFunction;
+    }
+    pub trait TypeLevelFunction {}
+}
+
+/// Type-level `enum` for pin function.
+pub trait Function: func_sealed::Function {}
+
+/// Value-level `enum` for pin function.
+#[allow(missing_docs)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum DynFunction {
+    Xip,
```

Comment:
pub enums should have comments on each variant, perhaps with a link to the datasheet?

Reply:
I couldn't find something to write on each that would add anything valuable.

This seems pretty redundant:
```rust
pub enum DynFunction {
    /// Value-level `variant` for the Xip [`Function`].
    Xip,
    /// Value-level `variant` for the Uart [`Function`].
    Uart,
    // […]
}
```

Reply:
Well you could write them in the case that people expect:

```rust
/// Describes the function currently assigned to a pin with a dynamic type.
///
/// A 'pin' on the RP2040 can be connected to different parts of the chip
/// internally - for example, it could be configured as a GPIO pin and connected
/// to the SIO block, or it could be configured as a UART pin and connected to
/// the UART block.
pub enum DynFunction {
    /// The 'XIP' (or Execute-in-place) function, which means talking to the QSPI Flash
    Xip,
    /// The 'UART' (or 'serial-port') function.
    Uart
}
```